src/binder/bind_node_visitor.cpp

```diff
@@ -0,0 +1,149 @@
+  // Save the upper level context
+  auto pre_context = context_;
+  context_ = std::make_shared<BinderContext>();
+  context_->upper_context = pre_context;
+
+  if (node->from_table != nullptr) node->from_table->Accept(this);
+  if (node->where_clause != nullptr) node->where_clause->Accept(this);
+  if (node->order != nullptr) node->order->Accept(this);
+  if (node->limit != nullptr) node->limit->Accept(this);
+  if (node->group_by != nullptr) node->group_by->Accept(this);
+  for (auto select_element : *(node->select_list)) {
+    select_element->Accept(this);
+  }
+
+  // Restore the upper level context
+  context_ = context_->upper_context;
+}
+
+// Some sub query nodes inside SelectStatement
+void BindNodeVisitor::Visit(const parser::JoinDefinition *node) {
+  // The columns in join condition can only bind to the join tables
+  node->left->Accept(this);
+  node->right->Accept(this);
+  node->condition->Accept(this);
+}
+
+void BindNodeVisitor::Visit(const parser::TableRef *node) {
+  // Nested select. Not supported in the current executors
+  if (node->select != nullptr) node->select->Accept(this);
+  // Join
+  else if (node->join != nullptr)
+    node->join->Accept(this);
+  // Multiple tables
+  else if (node->list != nullptr) {
+    for (parser::TableRef *table : *(node->list)) table->Accept(this);
+  }
+  // Single table
+  else {
+    context_->AddTable(node);
+  }
+}
+
+void BindNodeVisitor::Visit(const parser::GroupByDescription *node) {
+  for (auto col : *(node->columns)) {
+    col->Accept(this);
+  }
+  if (node->having != nullptr) node->having->Accept(this);
+}
+void BindNodeVisitor::Visit(const parser::OrderDescription *node) {
+  if (node->expr != nullptr) node->expr->Accept(this);
+}
+
+void BindNodeVisitor::Visit(const parser::UpdateStatement *node) {
+  context_ = std::make_shared<BinderContext>();
+
+  node->table->Accept(this);
+  if (node->where != nullptr) node->where->Accept(this);
+
+  // TODO: Update columns are not bound because they are char*
+  // not TupleValueExpression in update_statement.h
+
+  context_ = nullptr;
+}
+
+void BindNodeVisitor::Visit(const parser::DeleteStatement *node) {
+  context_ = std::make_shared<BinderContext>();
+
+  context_->AddTable(node->GetDatabaseName(), node->GetTableName());
+
+  if (node->expr != nullptr) node->expr->Accept(this);
+
+  context_ = nullptr;
+}
+
+void BindNodeVisitor::Visit(const parser::LimitDescription *) {}
+void BindNodeVisitor::Visit(const parser::CopyStatement *) {}
+void BindNodeVisitor::Visit(const parser::CreateStatement *) {}
+void BindNodeVisitor::Visit(const parser::InsertStatement *) {}
+void BindNodeVisitor::Visit(const parser::DropStatement *) {}
+void BindNodeVisitor::Visit(const parser::PrepareStatement *) {}
+void BindNodeVisitor::Visit(const parser::ExecuteStatement *) {}
+void BindNodeVisitor::Visit(const parser::TransactionStatement *) {}
+
+void BindNodeVisitor::Visit(expression::TupleValueExpression *expr) {
```

Comment:
No. Because the alias of a column is defined in the select clause, it should be invisible in other clauses of the same level. However, if the inner query defines a column alias, the outer query can see it but it is not supported now.